File: go/echo_bot.go
package wgrok

import (
	"context"
	"fmt"
	"net/http"
	"strings"

	wmh "github.com/3rg0n/webex-message-handler/go"
)

// WgrokEchoBot listens for echo messages, validates allowlist, strips prefix, relays back.
type WgrokEchoBot struct {
	config    *BotConfig
	allowlist *Allowlist
	logger    wmh.Logger
	handler   *wmh.WebexMessageHandler
	client    *http.Client
	cancel    context.CancelFunc
}

// NewEchoBot creates a new WgrokEchoBot.
func NewEchoBot(config *BotConfig) *WgrokEchoBot {
	return &WgrokEchoBot{
		config:    config,
		allowlist: NewAllowlist(config.Domains),
		logger:    GetLogger(config.Debug, "wgrok.echo_bot"),
		client:    &http.Client{},
	}
}

// Run connects to Webex and listens for echo messages. Blocks until ctx is cancelled or Stop is called.
func (b *WgrokEchoBot) Run(ctx context.Context) error {
	ctx, cancel := context.WithCancel(ctx)
	b.cancel = cancel

	h, err := wmh.New(wmh.Config{
		Token:  b.config.WebexToken,
		Logger: b.logger,
	})
	if err != nil {
		cancel()
		return fmt.Errorf("create handler: %w", err)
	}
	b.handler = h

	h.OnMessageCreated(func(msg wmh.DecryptedMessage) {
		b.onMessage(msg)
	})

	b.logger.Info("Echo bot starting")
	if err := h.Connect(ctx); err != nil {
		cancel()
		return fmt.Errorf("connect: %w", err)
	}
	b.logger.Info("Echo bot connected")

	<-ctx.Done()
	return ctx.Err()
}

// Stop disconnects the echo bot.
func (b *WgrokEchoBot) Stop(ctx context.Context) {
	if b.cancel != nil {
		b.cancel()
	}
	if b.handler != nil {
		_ = b.handler.Disconnect(ctx)
		b.handler = nil
	}
	b.logger.Info("Echo bot stopped")
}

// onMessageWithCards is used by tests to inject card data without HTTP fetches.
func (b *WgrokEchoBot) onMessageWithCards(msg wmh.DecryptedMessage, cards []interface{}) {
	sender := msg.PersonEmail
	text := strings.TrimSpace(msg.Text)

	if !b.allowlist.IsAllowed(sender) {
		b.logger.Warn(fmt.Sprintf("Rejected message from %s: not in allowlist", sender))
		return
	}

	if !IsEcho(text) {
		b.logger.Debug(fmt.Sprintf("Ignoring non-echo message from %s", sender))
		return
	}

	slug, payload, err := ParseEcho(text)
	if err != nil {
		b.logger.Error(fmt.Sprintf("Failed to parse echo message: %v", err))
		return
	}

	response := FormatResponse(slug, payload)

	if len(cards) > 0 {
		b.logger.Info(fmt.Sprintf("Relaying to %s: %s (with %d card(s))", sender, response, len(cards)))
		_, err = SendCard(b.config.WebexToken, sender, response, cards[0], b.client)
	} else {
		b.logger.Info(fmt.Sprintf("Relaying to %s: %s", sender, response))
		_, err = SendMessage(b.config.WebexToken, sender, response, b.client)
	}
	if err != nil {
		b.logger.Error(fmt.Sprintf("Failed to relay message: %v", err))
	}
}

func (b *WgrokEchoBot) onMessage(msg wmh.DecryptedMessage) {
	sender := msg.PersonEmail
	text := strings.TrimSpace(msg.Text)

	if !b.allowlist.IsAllowed(sender) {
		b.logger.Warn(fmt.Sprintf("Rejected message from %s: not in allowlist", sender))
		return
	}

	if !IsEcho(text) {
		b.logger.Debug(fmt.Sprintf("Ignoring non-echo message from %s", sender))
		return
	}

	slug, payload, err := ParseEcho(text)
	if err != nil {
		b.logger.Error(fmt.Sprintf("Failed to parse echo message: %v", err))
		return
	}

	response := FormatResponse(slug, payload)

	// Check for card attachments on the original message
	cards := b.fetchCards(msg.ID)

	if len(cards) > 0 {
		b.logger.Info(fmt.Sprintf("Relaying to %s: %s (with %d card(s))", sender, response, len(cards)))
		_, err = SendCard(b.config.WebexToken, sender, response, cards[0], b.client)
	} else {
		b.logger.Info(fmt.Sprintf("Relaying to %s: %s", sender, response))
		_, err = SendMessage(b.config.WebexToken, sender, response, b.client)
	}
	if err != nil {
		b.logger.Error(fmt.Sprintf("Failed to relay message: %v", err))
	}
}

func (b *WgrokEchoBot) fetchCards(messageID string) []interface{} {
	if messageID == "" {
		return nil
	}
	fullMsg, err := GetMessage(b.config.WebexToken, messageID, b.client)
	if err != nil {
		b.logger.Debug(fmt.Sprintf("Could not fetch message attachments: %v", err))
		return nil
	}
	return ExtractCards(fullMsg)
}

File: rust/src/echo_bot.rs
use reqwest::Client;
use serde_json::Value;
use tokio::sync::watch;
use webex_message_handler::{Config, DecryptedMessage, HandlerEvent, WebexMessageHandler};

use crate::allowlist::Allowlist;
use crate::config::BotConfig;
use crate::logging::{get_logger, WgrokLogger};
use crate::protocol::{format_response, is_echo, parse_echo};
use crate::webex;

pub struct WgrokEchoBot {
    config: BotConfig,
    allowlist: Allowlist,
    logger: WgrokLogger,
    client: Client,
}

impl WgrokEchoBot {
    pub fn new(config: BotConfig) -> Self {
        let logger = get_logger(config.debug, "wgrok.echo_bot");
        let allowlist = Allowlist::new(&config.domains);
        Self {
            config,
            allowlist,
            logger,
            client: Client::new(),
        }
    }

    pub async fn run(&self, mut shutdown_rx: watch::Receiver<bool>) -> Result<(), String> {
        let handler = WebexMessageHandler::new(Config {
            token: self.config.webex_token.clone(),
            ..Default::default()
        })
        .map_err(|e| format!("create handler: {}", e))?;

        let mut rx = handler
            .take_event_rx()
            .await
            .ok_or("failed to take event rx")?;

        handler
            .connect()
            .await
            .map_err(|e| format!("connect: {}", e))?;

        self.logger.info("Echo bot connected");

        loop {
            tokio::select! {
                event = rx.recv() => {
                    match event {
                        Some(HandlerEvent::MessageCreated(msg)) => {
                            let cards = self.fetch_cards(&msg.id).await;
                            self.on_message_with_cards(&msg, &cards).await;
                        }
                        None => break,
                        _ => {}
                    }
                }
                _ = shutdown_rx.changed() => {
                    break;
                }
            }
        }

        handler.disconnect().await;
        self.logger.info("Echo bot stopped");
        Ok(())
    }

    pub async fn on_message_with_cards(&self, msg: &DecryptedMessage, cards: &[Value]) {
        let sender = &msg.person_email;
        let text = msg.text.trim();

        if !self.allowlist.is_allowed(sender) {
            self.logger
                .warn(&format!("Rejected message from {}: not in allowlist", sender));
            return;
        }

        if !is_echo(text) {
            self.logger
                .debug(&format!("Ignoring non-echo message from {}", sender));
            return;
        }

        let (slug, payload) = match parse_echo(text) {
            Ok(v) => v,
            Err(e) => {
                self.logger
                    .error(&format!("Failed to parse echo message: {}", e));
                return;
            }
        };

        let response = format_response(&slug, &payload);

        let result = if !cards.is_empty() {
            self.logger.info(&format!(
                "Relaying to {}: {} (with {} card(s))",
                sender,
                response,
                cards.len()
            ));
            webex::send_card(
                &self.config.webex_token,
                sender,
                &response,
                &cards[0],
                &self.client,
            )
            .await
        } else {
            self.logger
                .info(&format!("Relaying to {}: {}", sender, response));
            webex::send_message(&self.config.webex_token, sender, &response, &self.client).await
        };

        if let Err(e) = result {
            self.logger
                .error(&format!("Failed to relay message: {}", e));
        }
    }

    async fn fetch_cards(&self, message_id: &str) -> Vec<Value> {
        if message_id.is_empty() {
            return vec![];
        }
        match webex::get_message(&self.config.webex_token, message_id, &self.client).await {
            Ok(msg) => webex::extract_cards(&msg),
            Err(e) => {
                self.logger
                    .debug(&format!("Could not fetch message attachments: {}", e));
                vec![]
            }
        }
    }
}
